Thanks and Happy Easter!

Now it almost works, but I see 2 different problem 

1) Not related to this PR but to the firstTForbit in general: running a workflow starting with `internal-dpl-clock`, e.g. like
![image](https://user-images.githubusercontent.com/7382029/113524114-c2cc1400-95ac-11eb-878d-c0170fb38ced.png)
I see that the `DataHeader::firstTForbit` and `tfCounter` are not propagated correctly from the `DPL/ENUM/XXX` message.
E.g.  by hacking the `its-digit-reader`, `its-clusterer` and `its-cooked-tracker` to show 
````
  auto dh = DataRefUtils::getHeader<o2::header::DataHeader*>(pc.inputs().getByPos(0));
  LOG(INFO) << "1stOrbit : " << dh->firstTForbit << " TFid: " << dh->tfCounter << " for " << dh->dataOrigin.str << "/" << dh->dataDescription.str;
````
I get the following:
````
o2-its-reco-workflow --disable-mc --run | tee xx.log
 grep '1stOrbit' xx.log
[1793329:its-digit-reader]: [01:18:21][INFO] 1stOrbit : 60 TFid: 0 for DPL/ENUM
[1793330:its-clusterer]: [01:18:21][INFO] 1stOrbit : 4294967295 TFid: 4294967295 for ITS/DIGITS
[1793332:its-cooked-tracker]: [01:18:22][INFO] 1stOrbit : 4294967295 TFid: 4294967295 for ITS/CLUSTERSROF
````

2) certainly related to this PR: if I run a workflow like
````
o2-raw-file-reader-workflow --detect-tf0 --input-conf raw/ITS/ITSraw.cfg | o2-itsmft-stf-decoder-workflow
````
![image](https://user-images.githubusercontent.com/7382029/113525813-57d40a80-95b7-11eb-9d57-2f4acdec2d5f.png)

I get an exception:
````
[1855511:its-stf-decoder]: [02:32:10][STATE] Starting FairMQ state machine --> IDLE
[1855511:its-stf-decoder]: [02:32:10][STATE] IDLE ---> INITIALIZING DEVICE
[1855511:its-stf-decoder]: [02:32:10][STATE] Exiting FairMQ state machine
[1855511:its-stf-decoder]: [02:32:10][ERROR] Unhandled std::exception reached the top of main: bad type for option: orbit-offset, device shutting down.
````

I don't understand why the `orbit-offset` option pops up in the `its-stf-decoder` device since it has its proper inputs.
The workflow somehow gets options screwed up: if I run only 1st part, it shows that the device has an options and 
````
o2-raw-file-reader-workflow --detect-tf0 --input-conf raw/ITS/ITSraw.cfg -h
...
Available data processors (full info with '--help full'):
  --internal-dpl-clock arg              No processor options
  --raw-file-reader arg                 Use '--help raw-file-reader' to display processor options
````
and the same with  ` -h raw-file-reader` shows as expected:
````
Data processor options: raw-file-reader:
  --orbit-offset-enumeration arg (=0)   1st injected orbit
  --orbit-multiplier-enumeration arg (=1) orbits/TForbit
  --start-value-enumeration arg (=0)    initial value for the enumeration
  --end-value-enumeration arg (=-1)     final value for the enumeration
  --step-value-enumeration arg (=1)     step between one value and the other
````

But the full piped workflow help shows:
````
o2-raw-file-reader-workflow --detect-tf0 --input-conf raw/ITS/ITSraw.cfg | o2-itsmft-stf-decoder-workflow  -h
...
Available data processors (full info with '--help full'):
  --internal-dpl-clock arg              No processor options
  --raw-file-reader arg                 No processor options
  --its-stf-decoder arg                 Use '--help its-stf-decoder' to display processor options
````
i.e. the options of my 1st device, `raw-file-reader` (which has not input thus gets your new options added), have disappeared.
And the help of `its-stf-decoder` (which produces this exception), does not show the option like `orbit-offset`, due to which the exception is produced...

I have the data needed for these workflows here: https://cernbox.cern.ch/index.php/s/DWoZdpTubY5yBDP

--- a/Framework/Core/include/Framework/ConfigParamRegistry.h
+++ b/Framework/Core/include/Framework/ConfigParamRegistry.h
@@ -83,6 +83,7 @@ class ConfigParamRegistry
     } catch (...) {
       throw std::invalid_argument(std::string("error parsing option: ") + key);
     }
+    throw std::invalid_argument(std::string("bad type for option: ") + key);
   }
 
  private:

--- a/Framework/Core/include/Framework/ConfigParamsHelper.h
+++ b/Framework/Core/include/Framework/ConfigParamsHelper.h
@@ -36,6 +36,10 @@ struct ConfigParamsHelper {
                                options_description& options,
                                boost::program_options::options_description vetos = options_description());
 
+  /// Add the ConfigParamSpec @a spec to @a specs if there is no parameter with
+  /// the same name already.
+  static void addOptionIfMissing(std::vector<ConfigParamSpec>& specs, ConfigParamSpec spec);
+
   /// populate boost program options for a complete workflow
   template <typename ContainerType>
   static boost::program_options::options_description

--- a/Framework/Core/include/Framework/DataDescriptorMatcher.h
+++ b/Framework/Core/include/Framework/DataDescriptorMatcher.h
@@ -41,6 +41,13 @@ struct ContextRef {
   inline bool operator==(ContextRef const& other) const;
 };
 
+/// Special positions for variables in context.
+enum ContextPos {
+  STARTTIME_POS = 0,    /// The DataProcessingHeader::startTime associated to the timeslice
+  TFCOUNTER_POS = 14,   /// The DataHeader::tfCounter associated to the timeslice
+  FIRSTTFORBIT_POS = 15 /// The DataHeader::firstTFOrbit associated to the timeslice
+};
+
 /// An element of the matching context. Context itself is really a vector of
 /// those. It's up to the matcher builder to build the vector in a suitable way.
 /// We do not have any float in the value, because AFAICT there is no need for

--- a/Framework/Core/include/Framework/ExpirationHandler.h
+++ b/Framework/Core/include/Framework/ExpirationHandler.h
@@ -8,17 +8,16 @@
 // granted to it by virtue of its status as an Intergovernmental Organization
 // or submit itself to any jurisdiction.
 
-#ifndef FRAMEWORK_EXPIRATIONHANDLER_H
-#define FRAMEWORK_EXPIRATIONHANDLER_H
+#ifndef O2_FRAMEWORK_EXPIRATIONHANDLER_H_
+#define O2_FRAMEWORK_EXPIRATIONHANDLER_H_
 
 #include "Framework/Lifetime.h"
 #include "Framework/RoutingIndices.h"
+#include "Framework/DataDescriptorMatcher.h"
 #include <cstdint>
 #include <functional>
 
-namespace o2
-{
-namespace framework
+namespace o2::framework
 {
 
 struct PartRef;
@@ -29,7 +28,7 @@ struct TimesliceSlot;
 struct ExpirationHandler {
   using Creator = std::function<TimesliceSlot(TimesliceIndex&)>;
   using Checker = std::function<bool(uint64_t timestamp)>;
-  using Handler = std::function<void(ServiceRegistry&, PartRef& expiredInput, uint64_t timestamp)>;
+  using Handler = std::function<void(ServiceRegistry&, PartRef& expiredInput, uint64_t timestamp, data_matcher::VariableContext& variables)>;
 
   RouteIndex routeIndex;
   Lifetime lifetime;
@@ -38,7 +37,6 @@ struct ExpirationHandler {
   Handler handler;
 };
 
-} // namespace framework
-} // namespace o2
+} // namespace o2::framework
 
-#endif // FRAMEWORK_EXPIRATIONHANDLER_H
+#endif // O2_FRAMEWORK_EXPIRATIONHANDLER_H_

--- a/Framework/Core/include/Framework/LifetimeHelpers.h
+++ b/Framework/Core/include/Framework/LifetimeHelpers.h
@@ -76,7 +76,8 @@ struct LifetimeHelpers {
   /// dataOrigin, dataDescrition and dataSpecification of the given @a route.
   /// The payload of each message will contain an incremental number for each
   /// message being created.
-  static ExpirationHandler::Handler enumerate(ConcreteDataMatcher const& spec, std::string const& sourceChannel);
+  static ExpirationHandler::Handler enumerate(ConcreteDataMatcher const& spec, std::string const& sourceChannel,
+                                              int64_t orbitOffset, int64_t orbitMultiplier);
 
   /// Create a dummy (empty) message every time a record expires, suing @a spec
   /// as content of the payload.

--- a/Framework/Core/include/Framework/TimesliceIndex.inc
+++ b/Framework/Core/include/Framework/TimesliceIndex.inc
@@ -122,7 +122,7 @@ inline uint32_t TimesliceIndex::getFirstTFOrbitForSlot(TimesliceSlot slot) const
 {
   assert(mVariables.size() > slot.index);
   // firstTForbit is always at register 15
-  auto pval = std::get_if<uint32_t>(&mVariables[slot.index].get(15));
+  auto pval = std::get_if<uint32_t>(&mVariables[slot.index].get(data_matcher::FIRSTTFORBIT_POS));
   if (pval == nullptr) {
     return -1;
   }
@@ -132,7 +132,7 @@ inline uint32_t TimesliceIndex::getFirstTFOrbitForSlot(TimesliceSlot slot) const
 inline uint32_t TimesliceIndex::getFirstTFCounterForSlot(TimesliceSlot slot) const
 {
   assert(mVariables.size() > slot.index);
-  // firstTForbit is always at register 15
+  // tfCounter is always at register 14
   auto pval = std::get_if<uint32_t>(&mVariables[slot.index].get(14));
   if (pval == nullptr) {
     return -1;

--- a/Framework/Core/src/ConfigParamsHelper.cxx
+++ b/Framework/Core/src/ConfigParamsHelper.cxx
@@ -89,6 +89,16 @@ void ConfigParamsHelper::populateBoostProgramOptions(
   }
 }
 
+void ConfigParamsHelper::addOptionIfMissing(std::vector<ConfigParamSpec>& specs, ConfigParamSpec spec)
+{
+  for (auto& old : specs) {
+    if (old.name == spec.name) {
+      return;
+    }
+  }
+  specs.push_back(spec);
+}
+
 /// populate boost program options making all options of type string
 /// this is used for filtering the command line argument
 bool ConfigParamsHelper::dpl2BoostOptions(const std::vector<ConfigParamSpec>& spec,

--- a/Framework/Core/src/DataDescriptorMatcher.cxx
+++ b/Framework/Core/src/DataDescriptorMatcher.cxx
@@ -106,9 +106,9 @@ bool StartTimeValueMatcher::match(header::DataHeader const& dh, DataProcessingHe
     }
     context.put({ref->index, dph.startTime / mScale});
     // We always put in 14 the tfCounter
-    context.put({14, dh.tfCounter});
+    context.put({TFCOUNTER_POS, dh.tfCounter});
     // We always put in 15 the firstTForbit
-    context.put({15, dh.firstTForbit});
+    context.put({FIRSTTFORBIT_POS, dh.firstTForbit});
     return true;
   } else if (auto v = std::get_if<uint64_t>(&mValue)) {
     return (dph.startTime / mScale) == *v;

--- a/Framework/Core/src/DataRelayer.cxx
+++ b/Framework/Core/src/DataRelayer.cxx
@@ -108,6 +108,7 @@ DataRelayer::ActivityStats DataRelayer::processDanglingInputs(std::vector<Expira
     }
     assert(mDistinctRoutesIndex.empty() == false);
     auto timestamp = mTimesliceIndex.getTimesliceForSlot(slot);
+    auto& variables = mTimesliceIndex.getVariablesForSlot(slot);
     // We iterate on all the hanlders checking if they need to be expired.
     for (size_t ei = 0; ei < expirationHandlers.size(); ++ei) {
       auto& expirator = expirationHandlers[ei];
@@ -137,7 +138,7 @@ DataRelayer::ActivityStats DataRelayer::processDanglingInputs(std::vector<Expira
       if (part.size() == 0) {
         part.parts.resize(1);
       }
-      expirator.handler(services, part[0], timestamp.value);
+      expirator.handler(services, part[0], timestamp.value, variables);
       activity.expiredSlots++;
 
       mTimesliceIndex.markAsDirty(slot, true);

--- a/Framework/Core/src/DeviceSpecHelpers.cxx
+++ b/Framework/Core/src/DeviceSpecHelpers.cxx
@@ -177,7 +177,10 @@ struct ExpirationHandlerHelpers {
       throw runtime_error("InputSpec for Timers must be fully qualified");
     }
     // We copy the matcher to avoid lifetime issues.
-    return [matcher = *m, sourceChannel](DeviceState&, ConfigParamRegistry const&) { return LifetimeHelpers::enumerate(matcher, sourceChannel); };
+    return [matcher = *m, sourceChannel](DeviceState&, ConfigParamRegistry const& config) {
+      // Timers do not have any orbit associated to them
+      return LifetimeHelpers::enumerate(matcher, sourceChannel, 0, 0);
+    };
   }
 
   static RouteConfigurator::ExpirationConfigurator expiringEnumerationConfigurator(InputSpec const& spec, std::string const& sourceChannel)
@@ -187,8 +190,10 @@ struct ExpirationHandlerHelpers {
       throw runtime_error("InputSpec for Enumeration must be fully qualified");
     }
     // We copy the matcher to avoid lifetime issues.
-    return [matcher = *m, sourceChannel](DeviceState&, ConfigParamRegistry const&) {
-      return LifetimeHelpers::enumerate(matcher, sourceChannel);
+    return [matcher = *m, sourceChannel](DeviceState&, ConfigParamRegistry const& config) {
+      size_t orbitOffset = config.get<int64_t>("orbit-offset-enumeration");
+      size_t orbitMultiplier = config.get<int64_t>("orbit-multiplier-enumeration");
+      return LifetimeHelpers::enumerate(matcher, sourceChannel, orbitOffset, orbitMultiplier);
     };
   }
 
@@ -222,17 +227,17 @@ struct ExpirationHandlerHelpers {
   {
     try {
       ConcreteDataMatcher concrete = DataSpecUtils::asConcreteDataMatcher(spec);
-      return [concrete, sourceChannel](DeviceState&, ConfigParamRegistry const&) {
+      return [concrete, sourceChannel](DeviceState&, ConfigParamRegistry const& config) {
         return LifetimeHelpers::dummy(concrete, sourceChannel);
       };
     } catch (...) {
       ConcreteDataTypeMatcher dataType = DataSpecUtils::asConcreteDataTypeMatcher(spec);
       ConcreteDataMatcher concrete{dataType.origin, dataType.description, 0xdeadbeef};
-      return [concrete, sourceChannel](DeviceState&, ConfigParamRegistry const&) {
+      return [concrete, sourceChannel](DeviceState&, ConfigParamRegistry const& config) {
         return LifetimeHelpers::dummy(concrete, sourceChannel);
       };
+      // We copy the matcher to avoid lifetime issues.
     }
-    // We copy the matcher to avoid lifetime issues.
   }
 };
 

--- a/Framework/Core/src/LifetimeHelpers.cxx
+++ b/Framework/Core/src/LifetimeHelpers.cxx
@@ -11,6 +11,7 @@
 #include "Framework/DataProcessingHeader.h"
 #include "Framework/InputSpec.h"
 #include "Framework/LifetimeHelpers.h"
+#include "Framework/DataDescriptorMatcher.h"
 #include "Framework/Logger.h"
 #include "Framework/RawDeviceService.h"
 #include "Framework/ServiceRegistry.h"
@@ -143,7 +144,7 @@ ExpirationHandler::Checker LifetimeHelpers::expireTimed(std::chrono::microsecond
 /// expires via this mechanism).
 ExpirationHandler::Handler LifetimeHelpers::doNothing()
 {
-  return [](ServiceRegistry&, PartRef& ref, uint64_t) -> void { return; };
+  return [](ServiceRegistry&, PartRef& ref, uint64_t, data_matcher::VariableContext&) -> void { return; };
 }
 
 // We simply put everything in a stringstream and read it afterwards.
@@ -188,7 +189,7 @@ ExpirationHandler::Handler
   if (matcher == nullptr) {
     throw runtime_error("InputSpec for Conditions must be fully qualified");
   }
-  return [spec, matcher, sourceChannel, serverUrl = prefix, overrideTimestampMilliseconds](ServiceRegistry& services, PartRef& ref, uint64_t timestamp) -> void {
+  return [spec, matcher, sourceChannel, serverUrl = prefix, overrideTimestampMilliseconds](ServiceRegistry& services, PartRef& ref, uint64_t timestamp, data_matcher::VariableContext&) -> void {
     // We should invoke the handler only once.
     assert(!ref.header);
     assert(!ref.payload);
@@ -265,7 +266,7 @@ ExpirationHandler::Handler
 /// FIXME: provide a way to customise the histogram from the configuration.
 ExpirationHandler::Handler LifetimeHelpers::fetchFromQARegistry()
 {
-  return [](ServiceRegistry&, PartRef& ref, uint64_t) -> void {
+  return [](ServiceRegistry&, PartRef& ref, uint64_t, data_matcher::VariableContext&) -> void {
     throw runtime_error("fetchFromQARegistry: Not yet implemented");
     return;
   };
@@ -276,18 +277,19 @@ ExpirationHandler::Handler LifetimeHelpers::fetchFromQARegistry()
 /// FIXME: provide a way to customise the histogram from the configuration.
 ExpirationHandler::Handler LifetimeHelpers::fetchFromObjectRegistry()
 {
-  return [](ServiceRegistry&, PartRef& ref, uint64_t) -> void {
+  return [](ServiceRegistry&, PartRef& ref, uint64_t, data_matcher::VariableContext&) -> void {
     throw runtime_error("fetchFromObjectRegistry: Not yet implemented");
     return;
   };
 }
 
 /// Enumerate entries on every invokation.
-ExpirationHandler::Handler LifetimeHelpers::enumerate(ConcreteDataMatcher const& matcher, std::string const& sourceChannel)
+ExpirationHandler::Handler LifetimeHelpers::enumerate(ConcreteDataMatcher const& matcher, std::string const& sourceChannel,
+                                                      int64_t orbitOffset, int64_t orbitMultiplier)
 {
   using counter_t = int64_t;
   auto counter = std::make_shared<counter_t>(0);
-  auto f = [matcher, counter, sourceChannel](ServiceRegistry& services, PartRef& ref, uint64_t timestamp) -> void {
+  return [matcher, counter, sourceChannel, orbitOffset, orbitMultiplier](ServiceRegistry& services, PartRef& ref, uint64_t timestamp, data_matcher::VariableContext& variables) -> void {
     // We should invoke the handler only once.
     assert(!ref.header);
     assert(!ref.payload);
@@ -299,6 +301,10 @@ ExpirationHandler::Handler LifetimeHelpers::enumerate(ConcreteDataMatcher const&
     dh.subSpecification = matcher.subSpec;
     dh.payloadSize = sizeof(counter_t);
     dh.payloadSerializationMethod = gSerializationMethodNone;
+    dh.tfCounter = timestamp;
+    dh.firstTForbit = timestamp * orbitMultiplier + orbitOffset;
+    variables.put({data_matcher::FIRSTTFORBIT_POS, dh.firstTForbit});
+    variables.put({data_matcher::TFCOUNTER_POS, dh.tfCounter});
 
     DataProcessingHeader dph{timestamp, 1};
 
@@ -312,7 +318,6 @@ ExpirationHandler::Handler LifetimeHelpers::enumerate(ConcreteDataMatcher const&
     ref.payload = std::move(payload);
     (*counter)++;
   };
-  return f;
 }
 
 /// Create a dummy message with the provided ConcreteDataMatcher
@@ -320,7 +325,7 @@ ExpirationHandler::Handler LifetimeHelpers::dummy(ConcreteDataMatcher const& mat
 {
   using counter_t = int64_t;
   auto counter = std::make_shared<counter_t>(0);
-  auto f = [matcher, counter, sourceChannel](ServiceRegistry& services, PartRef& ref, uint64_t timestamp) -> void {
+  auto f = [matcher, counter, sourceChannel](ServiceRegistry& services, PartRef& ref, uint64_t timestamp, data_matcher::VariableContext& variables) -> void {
     // We should invoke the handler only once.
     assert(!ref.header);
     assert(!ref.payload);
@@ -332,7 +337,23 @@ ExpirationHandler::Handler LifetimeHelpers::dummy(ConcreteDataMatcher const& mat
     dh.subSpecification = matcher.subSpec;
     dh.payloadSize = 0;
     dh.payloadSerializationMethod = gSerializationMethodNone;
-    dh.tfCounter = timestamp;
+
+    {
+      auto pval = std::get_if<uint32_t>(&variables.get(data_matcher::FIRSTTFORBIT_POS));
+      if (pval == nullptr) {
+        dh.firstTForbit = -1;
+      } else {
+        dh.firstTForbit = *pval;
+      }
+    }
+    {
+      auto pval = std::get_if<uint32_t>(&variables.get(data_matcher::TFCOUNTER_POS));
+      if (pval == nullptr) {
+        dh.tfCounter = timestamp;
+      } else {
+        dh.tfCounter = *pval;
+      }
+    }
 
     DataProcessingHeader dph{timestamp, 1};
 

--- a/Framework/Core/src/WorkflowHelpers.cxx
+++ b/Framework/Core/src/WorkflowHelpers.cxx
@@ -11,6 +11,7 @@
 #include "Framework/AlgorithmSpec.h"
 #include "Framework/AODReaderHelpers.h"
 #include "Framework/ChannelMatching.h"
+#include "Framework/ConfigParamsHelper.h"
 #include "Framework/CommonDataProcessors.h"
 #include "Framework/ConfigContext.h"
 #include "Framework/DeviceSpec.h"
@@ -263,6 +264,8 @@ void WorkflowHelpers::injectServiceDevices(WorkflowSpec& workflow, ConfigContext
     {ConfigParamSpec{"aod-file", VariantType::String, {"Input AOD file"}},
      ConfigParamSpec{"aod-reader-json", VariantType::String, {"json configuration file"}},
      ConfigParamSpec{"time-limit", VariantType::Int64, 0ll, {"Maximum run time limit in seconds"}},
+     ConfigParamSpec{"orbit-offset-enumeration", VariantType::Int64, 0ll, {"initial value for the orbit"}},
+     ConfigParamSpec{"orbit-multiplier-enumeration", VariantType::Int64, 0ll, {"multiplier to get the orbit from the counter"}},
      ConfigParamSpec{"start-value-enumeration", VariantType::Int64, 0ll, {"initial value for the enumeration"}},
      ConfigParamSpec{"end-value-enumeration", VariantType::Int64, -1ll, {"final value for the enumeration"}},
      ConfigParamSpec{"step-value-enumeration", VariantType::Int64, 1ll, {"step between one value and the other"}}},
@@ -289,6 +292,8 @@ void WorkflowHelpers::injectServiceDevices(WorkflowSpec& workflow, ConfigContext
     std::string prefix = "internal-dpl-";
     if (processor.inputs.empty() && processor.name.compare(0, prefix.size(), prefix) != 0) {
       processor.inputs.push_back(InputSpec{"enumeration", "DPL", "ENUM", static_cast<DataAllocator::SubSpecificationType>(compile_time_hash(processor.name.c_str())), Lifetime::Enumeration});
+      ConfigParamsHelper::addOptionIfMissing(processor.options, ConfigParamSpec{"orbit-offset-enumeration", VariantType::Int64, 0ll, {"1st injected orbit"}});
+      ConfigParamsHelper::addOptionIfMissing(processor.options, ConfigParamSpec{"orbit-multiplier-enumeration", VariantType::Int64, 0ll, {"orbits/TForbit"}});
       processor.options.push_back(ConfigParamSpec{"start-value-enumeration", VariantType::Int64, 0ll, {"initial value for the enumeration"}});
       processor.options.push_back(ConfigParamSpec{"end-value-enumeration", VariantType::Int64, -1ll, {"final value for the enumeration"}});
       processor.options.push_back(ConfigParamSpec{"step-value-enumeration", VariantType::Int64, 1ll, {"step between one value and the other"}});

--- a/Framework/Core/src/WorkflowSerializationHelpers.cxx
+++ b/Framework/Core/src/WorkflowSerializationHelpers.cxx
@@ -765,7 +765,7 @@ void WorkflowSerializationHelpers::dump(std::ostream& out,
     w.Key("options");
     w.StartArray();
     for (auto& option : processor.options) {
-      if (option.name == "start-value-enumeration" || option.name == "end-value-enumeration" || option.name == "step-value-enumeration") {
+      if (option.name == "start-value-enumeration" || option.name == "end-value-enumeration" || option.name == "step-value-enumeration" || option.name == "orbit-offset-enumeration" || option.name == "orbit-multiplier-enumeration") {
         continue;
       }
       w.StartObject();

